kep/399-sysched-scoring/README.md

```diff
@@ -0,0 +1,165 @@
+
+<p align="center">
+<img src="images/arch.png" style="width:70%">
+</p>
+
+<p align="center">
+<b>Figure 3: Overall architecture</b>
+</p>
+
+### Scheduling Metrics - Extraneous System Call (ExS)
+
+At the heart of SySched is a single-valued score used to quantify the extraneous system call exposure of a pod on a node. Informally, ExS captures the additional system calls accessible on a node that the pod itself does not use. This score reflects the potential “danger” a pod must face when running on the target node as vulnerabilities in these extraneous system calls from neighboring pods can jeopardize its own security. In the following figure, we can obtain the ExS for pod C1, C2, and C3 for a node with a kernel of nine system calls.
+
+<p align="center">
+<img src="images/exs.png" style="width:60%">
+</p>
+
+<p align="center">
+<b>Figure 4: Example ExS score calculation</b>
+</p>
+
+### Scheduling Plugin
+
+We developed a new plugin to implement the scoring extension API point in the Kubernetes scheduler while leaving other filters and scoring plugins intact. We leverage existing filtering and scoring operations of the Kubernetes scheduler to handle other aspects of pod placement such as spreading the pods and ensuring resource availability.
+
+ There are two main components of our plugin: 1) a lightweight in memory state store that holds the up-to-date node to pod mapping and 2) mechanisms for retrieving the system call sets used by pods and for computing the ExS scores. The state store is implemented as a thread that listens to events of a pod’s lifecycle, e.g., creation, stop, destruction, run, etc. Using these event notifications, this thread tracks where pods are currently running in a cluster. When a pod stops running on a node, our plugin also removes it from its internal mapping. While the entire state of pod placement in a cluster can be retrieved dynamically by querying the API server, this can be an expensive task to do per pod scheduling event, hence we choose to maintain this state internally in our scheduler.
+
+### System call Profiles
+
+**How the Scheduler gets access to pods' system call profiles**
+
+We assume that the system call profiles of pods have explicitly been made available to the scheduler through a pod's security context. The seccomp profile directory on the host is mounted into the scheduler's pod. Then it reads and parses the seccomp profiles and obtains the the system call set for that pod. If the Security Profile Operator (SPO) is used (see below), the scheduler can instead use API server to access the seccomp profile CRD to get the system call profile. With the SPO, there's no need to mount the host's seccomp profile directory into the scheduler pod.
```

Comment:
> The seccomp profile directory on the host is mounted into the scheduler's pod.

Can you elaborate? How the scheduler (usually deployed in the control plane) access the seccomp profiles specified in a to-schedule pod? The seccomp profile settings are specific to a particular node which is not determined at the scheduling time yet.

Reply:
The assumption is that we would have seccomp profiles for all the to-be-deployed pods out of band. This can be either using the SPO or just the developers/users generating the profiles for their pods and use cases. The seccomp profiles are then collected and mounted as a volume in the scheduler's pod. This works but not very flexible as using the CRD from the SPO. We will make changes to better clarify in the KEP. We might make this option a minor alternative to using the SPO as that's a better way of doing things.

Reply:
Reading from the SPO CRD makes sense to me. It's just I don't think reading and mounting the regular seccomp profile(s) is an option.

Reply:
Agreed. We've updated the KEP to reflect only the SPO option and have removed the mounting option.